website/pages/docs/job-specification/hcl2/variables.mdx

```diff
@@ -0,0 +1,307 @@
+
+Each input variable accepted by a job must be declared using a `variable`
+block :
+
+```hcl
+variable "image_id" {
+  type = string
+}
+
+variable "availability_zone_names" {
+  type    = list(string)
+  default = ["us-west-1a"]
+}
+
+variable "docker_ports" {
+  type = list(object({
+    internal = number
+    external = number
+    protocol = string
+  }))
+  default = [
+    {
+      internal = 8300
+      external = 8300
+      protocol = "tcp"
+    }
+  ]
+}
+```
+
+Or a less precise variables block:
+
+```hcl
+variables {
+  foo       = "value"
+  my_secret = "foo"
+}
+```
+
+The label after the `variable` keyword or a label of a `variables` block is a
+name for the variable, which must be unique among all variables in the same
+job. This name is used to assign a value to the variable from outside and to
+reference the variable's value from within the job.
+
+The `variable` block can optionally include a `type` argument to specify what
+value types are accepted for the variable, as described in the following
+section.
+
+The `variable` declaration can also include a `default` argument. If present,
+the variable is considered to be _optional_ and the default value will be used
+if no value is set when calling the job or running Nomad. The `default`
+argument requires a literal value and cannot reference other objects in the
+configuration.
+
+## Using Input Variable Values
+
+Within the job that declared a variable, its value can be accessed from
+within [expressions](/docs/job-specification/hcl2/expressions) as `var.<NAME>`, where `<NAME>`
+matches the label given in the declaration block:
+
+```hcl
+config {
+  image = var.task_image
+  label = var.task_labels
+}
+```
+
+The value assigned to a variable can be accessed only from expressions within
+the folder where it was declared.
+
+## Type Constraints
+
+The `type` argument in a `variable` block allows you to restrict the [type of
+value](/docs/job-specification/hcl2/expressions#types-and-values) that will be accepted as the value
+for a variable. If no type constraint is set then a value of any type is
+accepted.
+
+While type constraints are optional, we recommend specifying them; they serve
+as easy reminders for users of the job, and allow Nomad to return a helpful
+error message if the wrong type is used.
+
+Type constraints are created from a mixture of type keywords and type
+constructors. The supported type keywords are:
+
+- `string`
+- `number`
+- `bool`
+
+The type constructors allow you to specify complex types such as collections:
+
+- `list(<TYPE>)`
+- `set(<TYPE>)`
+- `map(<TYPE>)`
+- `object({<ATTR NAME> = <TYPE>, ... })`
+- `tuple([<TYPE>, ...])`
+
+The keyword `any` may be used to indicate that any type is acceptable. For more
+information on the meaning and behavior of these different types, as well as
+detailed information about automatic conversion of complex types, see [Type
+Constraints](https://www.terraform.io/docs/configuration/types.html).
```

Comment:
Was this intended to be pointing to the Terraform docs?

Reply:
Yeah, it didn't quite fit in here in our hierarchy and the page didn't have any Terraform info, so I left it as-is.  However, I'll bring it in in my next iteration to avoid having the reader leave the Nomad site.